packages/flutter/lib/src/material_animated_icons/animated_icons.dart

```diff
@@ -0,0 +1,269 @@
+    final TextDirection textDirection = this.textDirection ?? Directionality.of(context);
+    return new Semantics(
+      label: semanticLabel,
+      child: new CustomPaint(
+        size: new Size(iconSize, iconSize),
+        painter: new _AnimatedIconPainter(
+          iconData.paths,
+          progress,
+          color ?? iconTheme.color,
+          iconSize / iconData.size.bottomRight(const Offset(0.0, 0.0)).dx,
+          textDirection == TextDirection.rtl,
+          _pathFactory,
+        ),
+      ),
+    );
+  }
+}
+
+typedef ui.Path _UiPathFactory();
+
+class _AnimatedIconPainter extends CustomPainter {
+  _AnimatedIconPainter(
+    this.paths,
+    this.progress,
+    this.color,
+    this.scale,
+    this.shouldMirror,
+    this.uiPathFactory,
+  ) : super(repaint: progress);
+
+  // This list is assumed to be immutable, changes to the contents of the list
+  // will not trigger a redraw as shouldRepaint will keep returning false.
+  final List<_PathFrames> paths;
+  final Animation<double> progress;
+  final Color color;
+  final double scale;
+  /// If this is true the image will be mirrored horizontally.
+  final bool shouldMirror;
+  final _UiPathFactory uiPathFactory;
+
+  @override
+  void paint(ui.Canvas canvas, Size size) {
+    canvas.scale(scale, scale);
+    if (shouldMirror) {
+      canvas.rotate(math.pi);
+      canvas.translate(-size.width, -size.height);
+    }
+
+    for (_PathFrames path in paths)
+      path.paint(canvas, color, uiPathFactory, progress.value.clamp(0.0, 1.0));
+  }
+
+
+  @override
+  bool shouldRepaint(_AnimatedIconPainter oldDelegate) {
+    return oldDelegate.progress.value != progress.value
+      || oldDelegate.color != color
+      // We are comparing the paths list by reference, assuming the list is
+      // treated as immutable to be more efficient.
+      || oldDelegate.paths != paths
+      || oldDelegate.scale != scale
+      || oldDelegate.uiPathFactory != uiPathFactory;
+  }
+
+  @override
+  bool hitTest(Offset position) => null;
+
+  @override
+  bool shouldRebuildSemantics(CustomPainter oldDelegate) => false;
+
+  @override
+  SemanticsBuilderCallback get semanticsBuilder => null;
+}
+
+class _PathFrames {
+  const _PathFrames({
+    @required this.commands,
+    @required this.opacities
+  });
+
+  final List<_PathCommand> commands;
+  final List<double> opacities;
+
+  void paint(ui.Canvas canvas, Color color, _UiPathFactory uiPathFactory, double progress) {
+    final double opacity = _interpolate(opacities, progress, lerpDouble);
+    final ui.Paint paint = new ui.Paint()
+      ..style = PaintingStyle.fill
+      ..color = color.withOpacity(opacity);
+    final ui.Path path = uiPathFactory();
+    for (_PathCommand command in commands)
+      command.apply(path, progress);
+    canvas.drawPath(path, paint);
+  }
+}
+
+abstract class _PathCommand {
+  const _PathCommand();
+
+  void apply(ui.Path path, double progress);
+}
```

Comment:
this class could do with some documentation explaining what's going on

Reply:
done